File: tutorial_code/chapter_1/colour_tracking.cpp
#include "stdio.h"
#include "opencv2/core/core.hpp"
#include "opencv2/core/core_c.h"
#include "opencv2/highgui/highgui_c.h"
#include "opencv2/imgproc/imgproc_c.h"

using namespace cv;
using namespace std;

int lowerH=0;
int lowerS=0;
int lowerV=0;

int upperH=180;
int upperS=256;
int upperV=256;

IplImage* imgThresh;
IplImage* imgHSV;
IplImage* frame;

//This function threshold the HSV image and create a binary image
IplImage* GetThresholdedImage(IplImage* imgHSV){

	IplImage* imgThresh=cvCreateImage(cvGetSize(imgHSV),IPL_DEPTH_8U, 1);
	cvInRangeS(imgHSV, cvScalar(lowerH,lowerS,lowerV), cvScalar(upperH,upperS,upperV), imgThresh); 

	return imgThresh;
}

//This function create two windows and 6 trackbars for the "Settings" window
void setwindowSettings(){
	cvNamedWindow("Filtered", 2);
	cvNamedWindow("Original", 2);
	cvNamedWindow("Settings", 2);

	cvCreateTrackbar("LowerH", "Settings", &lowerH, 180, NULL);
	cvCreateTrackbar("UpperH", "Settings", &upperH, 180, NULL);

	cvCreateTrackbar("LowerS", "Settings", &lowerS, 256, NULL);
	cvCreateTrackbar("UpperS", "Settings", &upperS, 256, NULL);

	cvCreateTrackbar("LowerV", "Settings", &lowerV, 256, NULL);
	cvCreateTrackbar("UpperV", "Settings", &upperV, 256, NULL); 
}

int main(int argc, const char **argv){

	if (argc == 1) {
		fprintf(stderr, "Usage: colour_tracking [File]");
		return 0;
	}

	frame = cvLoadImage(argv[1], CV_LOAD_IMAGE_COLOR);
	if (!frame) return 0;
	
	setwindowSettings();
	//iterate through each frames of the video
	while(true){
	
		imgHSV = cvCreateImage(cvGetSize(frame), IPL_DEPTH_8U, 3); 
		cvCvtColor(frame, imgHSV, CV_BGR2HSV); //Change the color format from BGR to HSV

		imgThresh = GetThresholdedImage(imgHSV);

		cvShowImage("Filtered", imgThresh);
		cvShowImage("Original", frame);

		//Clean up used images
		cvReleaseImage(&imgHSV);
		cvReleaseImage(&imgThresh);

		//Wait 80mS
		int c = cvWaitKey(80);
		//If 'ESC' is pressed, break the loop
		if((char)c==27 ) break;

	}

	cvDestroyAllWindows();

	return 0;
}

File: tutorial_code/chapter_1/colour_tracking.py
# initialise Original capture
import cv2
import numpy as np
import sys


def do_nothing(*arg):
    # print arg
    pass

def setwindowSettings():
    # initial values
    lowerH = 0
    lowerS = 0
    lowerV = 0
    upperH = 180
    upperS = 256
    upperV = 256
    cv2.namedWindow("Settings", 2)
    cv2.namedWindow("Original", 2)
    cv2.namedWindow("Filtered", 2)
    cv2.createTrackbar('LowerH', "Settings", lowerH, 180, do_nothing)
    cv2.createTrackbar('UpperH', "Settings", upperH, 180, do_nothing)
    cv2.createTrackbar('LowerS', "Settings", lowerS, 256, do_nothing)
    cv2.createTrackbar('UpperS', "Settings", upperS, 256, do_nothing)
    cv2.createTrackbar('LowerV', "Settings", lowerV, 256, do_nothing)
    cv2.createTrackbar('UpperV', "Settings", upperV, 256, do_nothing)


def getThreshImage(image):
        
    imgHSV = cv2.cvtColor(image, cv2.cv.CV_RGB2HSV)
    lowerH = cv2.getTrackbarPos('LowerH', 'Settings')
    lowerS = cv2.getTrackbarPos('LowerS', 'Settings')
    lowerV = cv2.getTrackbarPos('LowerV', 'Settings')
    
    upperH = cv2.getTrackbarPos('UpperH', 'Settings')
    upperS = cv2.getTrackbarPos('UpperS', 'Settings')
    upperV = cv2.getTrackbarPos('UpperV', 'Settings')
    result = cv2.inRange(imgHSV, np.array((lowerH, lowerS, lowerV)), np.array((upperH, upperS, upperV)))
    
    # print time.time() - t1
    return result
    # return cv2.inRange(imgHSV, np.array((lowerH, lowerS, lowerV)), np.array((upperH, upperS, upperV)))


def main():

    # return if no file path given
    if (len(sys.argv) == 1):
        return

    setwindowSettings()

    while(1):
        frame = cv2.imread(sys.argv[1])
        cv2.imshow("Filtered", getThreshImage(frame))
        cv2.imshow("Original", frame)

        c = cv2.waitKey(80)
        if c == 27:
            break
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()